### PR TITLE
Denormalization tutorial

### PR DESCRIPTION
**Pull request by ekmb**

### CODE REVIEW

File: tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "desirable-words",
+   "metadata": {},
+   "source": [
+    "# NeMo Inverse Text Normalization\n",
+    "\n",
+    "The NeMo ITN tool is a Python package that is based on weighted finite-state\n",
+    "transducer (WFST) grammars. The tool uses [`Pynini`](https://github.com/kylebgorman/pynini) to construct WFSTs, and the created grammars can be exported and integrated into [`Sparrowhawk`](https://github.com/google/sparrowhawk) (an open source version of [The Kestrel TTS text normalization system](https://www.cambridge.org/core/journals/natural-language-engineering/article/abs/kestrel-tts-text-normalization-system/F0C18A3F596B75D83B75C479E23795DA)) for production. The NeMo ITN tool is a Python extension of `Sparrowhawk`. \n",
```
Comment (yzhang123) [inline tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb:44]: NeMo ITN tool `can be seen` a Python extension of `Sparrowhawk

File: tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "Currently, NeMo tool provides support for English language and the following semiotic classes from the [Google Text normalization dataset](https://www.kaggle.com/richardwilliamsproat/text-normalization-for-english-russian-and-polish):\n",
```
Comment (yzhang123) [inline tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb:46]: for `English and` [..]

File: tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "limiting-genesis",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from nemo_tools.text_denormalization.denormalize import denormalize\n",
```
Comment (yzhang123) [inline tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb:72]: emphasize that ITN is part of package nemo_tools, which can be installed as shown in main README

**Review by yzhang123 — COMMENTED** [review]
added comments, structure looks good

File: tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "Although a large fraction of this dataset can be reused for ITN by swapping input with output, the dataset is not bijective. \n",
+    "\n",
+    "For example: `1,000 -> one thousand`, `1000 -> one thousand`, `3:00pm -> three p m`, `3 pm -> three p m` are valid data samples for normalization but the inverse does not hold for ITN. \n",
+    "\n",
+    "We used regex rules to disambiguate samples where possible, see `nemo_tools/text_denormalization/clean_eval_data.py`.\n",
+    "\n",
+    "To run evaluation, the `input` file should follow the Google Text normalization dataset format.\n",
+    "\n",
+    "Example evaluation run: python run_evaluate.py --input=./en_with_types/output-00001-of-00100 [--denormalizer nemo]"
```
[the remark this answers is not in the document]
Comment (ekmb) [reply tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb]: done

File: tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "adult-elephant",
+   "metadata": {},
+   "source": [
+    "Here `0` - is a start note, `1` and `2` are the accept nodes, while `3` is a finite state.\n",
+    "By default (token_type=\"byte\", `Pynini` interprets the string as a sequence of bytes, assigning one byte per arc. \n",
+    "\n",
+    "A finite state transducer (FST) not only matches the pattern but also produces output according to the defined transition states."
```
[the remark this answers is not in the document]
Comment (ekmb) [reply tutorials/tools/Denormalizaion-Inverse_Text_Normalization.ipynb]: done

File: tutorials/tools/Text_Denormalization_(Inverse_Text_Normalization).ipynb [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "twenty-terrorist",
+   "metadata": {},
+   "source": [
+    "NeMo ITN defines the following APIs that are called in sequence:\n",
+    "\n",
+    "- `classify()` - creates a linear automaton from the input string and composes it with the final classification WFST, which transduces numbers and inserts semantic tags.  \n",
```
Comment (yzhang123) [inline tutorials/tools/Text_Denormalization_(Inverse_Text_Normalization).ipynb]: they have different names now, could you use signature in current ode?
classify i think is find_tag + select_tag,
verbalize is find_veerbalize + select_verbalize
etc

File: docs/source/nemo_tools/text_denormalization.rst [hunk 7]
```diff
@@ -0,0 +1,13 @@
+Text Denormalization
+====================
+
+Inverse text normalization (ITN), also called denormalization, is a part of the Automatic Speech Recognition (ASR) post-processing pipeline.
+ITN is the task of converting the raw spoken output of the ASR model into its written form to improve the text readability.
```
Comment (yzhang123) [inline docs/source/nemo_tools/text_denormalization.rst:5]: Could you do this in new sentence with --> arrows so the input -> output stands more out?

**Review by yzhang123 — CHANGES_REQUESTED** [review]

**Review by yzhang123 — APPROVED** [review]

**Pull request merged**